### ISSUE TITLE
firecracker-0.21.x pre-boot hang?

### ISSUE THREAD

**Issue by thepwagner** [firecracker-microvm/firecracker#1674]
We're running `0.20` builds of firecracker+jailer, using the API server via https://github.com/firecracker-microvm/firecracker-go-sdk.

While evaluating the `0.21.0` and `0.21.1` releases, we found occasionally (~1/1000 VMs) firecracker will block with 100% CPU usage, and not actually boot the guest.

VMM logs of hung boots were identical to VMs that booted without issue.
`gdb` did not seem productive: two threads in `epoll_wait` and the vmm_cpu thread in `ioctl`.
Is there any other debugging context I can provide?

This isn't an issue with `0.20` and I couldn't reproduce it with a `devtool build` from `master`. 
This may be a wontfix/wait for `0.22`.

**Comment by dianpopa** [comment]
Hi @thepwagner ,

Sorry to hear you stumbled upon trouble. I will try to replicate this issue and dig into the problem. I understand the API socket does not appear for the hanging microVM so when you say "VMM logs",  are you referring to the dmesg logs?

**Comment by thepwagner** [comment]
:wave: @dianpopa 

The API socket appears: the client is able to connect, configure and "start" the machine.
But the machine doesn't produce any output (e.g. the guest kernel bootstrapping) - it just sits at 100% CPU until killed.

By VMM logs I mean https://github.com/firecracker-microvm/firecracker-go-sdk/blob/e6eaff81c885f51574fd7c1f7163c05e03c2ca52/machine.go#L609 / https://github.com/firecracker-microvm/firecracker/blob/v0.21.1/src/api_server/src/request/logger.rs#L9 

I hadn't checked `dmesg`; will do!

**Comment by thepwagner** [comment]
Logs from a hanging process:
```
Running Firecracker v0.21.1
2020-03-13T18:12:39.559859029 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.563777938 [534:INFO] The API server received a Put request on "/machine-config" with body "{\"cpu_template\":\"T2\",\"ht_enabled\":false,\"mem_size_mib\":512,\"vcpu_count\":1}\n".
2020-03-13T18:12:39.563821260 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.564676967 [534:INFO] The API server received a Get request on "/machine-config".
2020-03-13T18:12:39.564712129 [534:INFO] The request was executed successfully. Status code: 200 OK.
2020-03-13T18:12:39.568337264 [534:INFO] The API server received a Put request on "/boot-source" with body "{\"boot_args\":\"pci=off nomodules quiet ro noapic reboot=k panic=1\",\"kernel_image_path\":\"44ed6229557ca6efd1846d2236aa2742bddcf143ecc3434e94239465ee632788\"}\n".
2020-03-13T18:12:39.568412102 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.575779725 [534:INFO] The API server received a Put request on "/drives/0" with body "{\"drive_id\":\"0\",\"is_read_only\":true,\"is_root_device\":false,\"path_on_host\":\"config-984051412.ext4\"}\n".
2020-03-13T18:12:39.575848161 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.583783595 [534:INFO] The API server received a Put request on "/drives/1" with body "{\"drive_id\":\"1\",\"is_read_only\":true,\"is_root_device\":false,\"path_on_host\":\"2c5bd5c597790e612b02b0004609f13cd4c60c83.ext4\"}\n".
2020-03-13T18:12:39.583849788 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.595783739 [534:INFO] The API server received a Put request on "/drives/2" with body "{\"drive_id\":\"2\",\"is_read_only\":false,\"is_root_device\":false,\"path_on_host\":\"tmp-681462054\"}\n".
2020-03-13T18:12:39.595855841 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.601608827 [534:INFO] The API server received a Put request on "/drives/root_drive" with body "{\"drive_id\":\"root_drive\",\"is_read_only\":false,\"is_root_device\":true,\"path_on_host\":\"tmp-680929799\"}\n".
2020-03-13T18:12:39.601671491 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.606535827 [534:INFO] The API server received a Put request on "/network-interfaces/1" with body "{\"guest_mac\":\"00:03:ff:a7:4e:3d\",\"host_dev_name\":\"tap0\",\"iface_id\":\"1\"}\n".
2020-03-13T18:12:39.606898110 [534:INFO] The request was executed successfully. Status code: 204 No Content.
2020-03-13T18:12:39.607276037 [534:INFO] The API server received a Put request on "/actions" with body "{\"action_type\":\"InstanceStart\"}\n".
2020-03-13T18:12:39.607314544 [534:INFO] VMM received instance start command
2020-03-13T18:12:39.607337972 [534:INFO] Guest memory starts at 0x7f463b0fb000
2020-03-13T18:12:39.658529747 [534:INFO] The request was executed successfully. Status code: 204 No Content.
```

Burning CPU:
```
$ for i in `seq 10`; do ps aux | grep "firecracker --id 534" | grep -v grep | sed 's/^[^ ]* /USER /g'; sleep 1; done
USER 63399 98.9  0.0 530232 33784 ?        Sl   18:12   7:14 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 98.9  0.0 530232 33784 ?        Sl   18:12   7:15 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 98.9  0.0 530232 33784 ?        Sl   18:12   7:16 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 99.0  0.0 530232 33784 ?        Sl   18:12   7:17 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 99.0  0.0 530232 33784 ?        Sl   18:12   7:18 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 99.0  0.0 530232 33784 ?        Sl   18:12   7:19 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 99.0  0.0 530232 33784 ?        Sl   18:12   7:20 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 99.0  0.0 530232 33784 ?        Sl   18:12   7:21 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 98.8  0.0 530232 33784 ?        Sl   18:12   7:23 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
USER 63399 98.9  0.0 530232 33784 ?        Sl   18:12   7:24 /firecracker --id 534 --start-time-us 491973889 --start-time-cpu-us 644 --seccomp-level 2
```

```
$ rust-gdb -p 63399
GNU gdb (Debian 8.2.1-2+b3) 8.2.1
...
Attaching to process 63399
[New LWP 63507]
[New LWP 63603]
0x000000000050f0a7 in epoll_pwait ()
(gdb) info threads
  Id   Target Id               Frame 
* 1    LWP 63399 "firecracker" 0x000000000050f0a7 in epoll_pwait ()
  2    LWP 63507 "fc_api"      0x000000000050f0a7 in epoll_pwait ()
  3    LWP 63603 "fc_vcpu 0"   0x0000000000510690 in ioctl ()
```

Dmesg seems uninteresting - unrelated messages from a nearby `containerd`:
```
$ dmesg | wc -l
234
$ dmesg | grep "tap0: link becomes ready" | wc -l
191
$ dmesg | grep "overlayfs" | wc -l
43
```

Our test suite queues 5000 VMs that sleep 5-10 seconds before self-terminating.
The runner is restricted to 200 concurrent guest VMs, but there is no rate limiting.
From this pass, 2 of the 5000 jobs ended up in this state.

**Comment by dianpopa** [comment]
@thepwagner unfortunately I was not able to replicate the behavior. What I did though was go through the commits between 0.20.0 and 0.21.0 and try to spot that/those commits that could trigger the behavior you describe. Could you please test this [branch](https://github.com/dianpopa/firecracker/commits/i1674) I created for you and let me know if the behavior still replicates? The branch mirrors v0.21.0 without commits 3a9a1ac, 776ccda, 646b83b. 

Looking forward to the result!

**Comment by thepwagner** [comment]
@dianpopa thanks for the pointers!
I had to fixup `src/vmm/src/vstate.rs:1084` but otherwise tested that branch - I still ended up with 1 of 1000 VMs spinning as previously described.

I built `firecracker`+`jailer` from 6a4bb667313956b353283681fd6fc02f95d55cdc (`master`), re-ran without changes to our client and repeated with 0 of 5000 VMs stuck.
As things behave fine on `master`, I'm not concerned so will close this issue.

I kept back testing and ended up reproducing this all the way back to `0.17.0`; it may be caused by something else in our environment changing around the same time as we attempted the `0.21.0` upgrade.
All my testing was on `4.19.0-8-cloud-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26)` / debian-buster (`ami-04d70e069399af2e9` on us-east-1).

Thanks for taking a look, sorry for the red herring.
